scylla-gdb.py

```diff
@@ -603,12 +570,29 @@ def __init__(self, ref):
         else:
             self.buf_size = 1
 
+    def _foreach_node(self):
+        node_it = self.ref['_M_impl']['_M_start']
+        node_end = self.ref['_M_impl']['_M_finish']
+        node = node_it['_M_node']
+        it = node_it['_M_first']
+        end = node_it['_M_last']
+
+        if not bool(node):
+            return
+
+        while node != node_end['_M_node']:
+            yield it, end
+            node = node + 1
+            it = node.dereference()
+            end = it + self.buf_size
+
+        yield node_end['_M_first'], node_end['_M_cur']
+
     def __len__(self):
-        start = self.ref['_M_impl']['_M_start']
-        finish = self.ref['_M_impl']['_M_finish']
-        return (self.buf_size * (max(1, finish['_M_node'] - start['_M_node']) - 1) +
-            start['_M_last'] - start['_M_cur'] +
-            finish['_M_cur'] - finish['_M_first'])
+        l = 0
+        for start, end in self._foreach_node():
+            l += (end - start)
+        return l
```

Comment:
This can't be the most efficient way to calculate the correct size of an std::dequeue, because officially std::dequeue::size() has O(1) complexity and this calculation doesn't. I don't understand the old code or where it has an error, but there should be a way to calculate the length correctly without a loop.
I'm not saying this is important (it's just gdb, not a high-performance run) - maybe the simplicity of your implementation is more important than the performance (and incorrectness...) of the old implementation. But I still want to point this out.

Reply:
The previous implementation was O(1) and it was incorrect. I could probably have fixed it but preferred instead to go for the simple thing and reuse the same node-iteration across iteration and length calculation. Note that this is amortized O(1) as it iterates over nodes, each containing 512 elements. So this is `O(n/512)`.

Reply:
O(n/512) is not "amortized O(1)", it's O(n), but never mind, I agree with your point - for debugging purposes it's better to have a correct implementation than a tiny-bit-more-efficient one (especially considering that std::dequeue is pretty crappy in its memory allocation behavior for large Ns anyway, so we don't expect N to be huge.).

Reply:
Yes you are right, I started out convincing you that the new `__len__()` is not `O(n)`, and by the end I realized this is not really true but didn't go back to fix the start of my paragraph.